--- a/external/dcap_source/QuoteGeneration/quote_wrapper/quote/linux/Makefile	2021-01-11 18:09:47.163528235 +0000
+++ b/external/dcap_source/QuoteGeneration/quote_wrapper/quote/linux/Makefile	2021-01-11 18:10:09.371240187 +0000
@@ -28,6 +28,7 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #
 #
+.NOTPARALLEL: all
 
 ######## SGX SDK Settings ########
 TOP_DIR  = ../../..
@@ -47,7 +48,7 @@
 Quote_C_Files := qe3_u.c se_thread.c se_trace.c
 Quote_Cpp_Files := $(wildcard ../*.cpp)
 
-Quote_Include_Paths := -I$(SGX_SDK)/include -I../inc -I../../common/inc -I./ -I../enclave -I$(TOP_DIR)/pce_wrapper/inc -I$(COMMON_DIR)/inc/internal -I$(COMMON_DIR)/inc/internal/linux
+Quote_Include_Paths := -I$(SGX_SDK)/include -I../inc -I../../common/inc -I. -I../enclave -I$(TOP_DIR)/pce_wrapper/inc -I$(COMMON_DIR)/inc/internal -I$(COMMON_DIR)/inc/internal/linux
 
 Quote_C_Flags := $(COMMON_FLAGS) -g -fPIC -Wno-attributes $(Quote_Include_Paths)
 
@@ -70,6 +70,9 @@
 .PHONY: all
 all:  install_lib
 
+$(BUILD_DIR):
+	mkdir -p $(BUILD_DIR)
+
 install_lib: $(Quote_Name) | $(BUILD_DIR)
 	@$(CP) $(Quote_Name) $|
 
@@ -78,7 +82,11 @@
 qe3_u.c: $(SGX_EDGER8R) ../enclave/qe3.edl
 	$(SGX_EDGER8R) --untrusted ../enclave/qe3.edl --search-path ../enclave
 	@echo "GEN  =>  $@"
 
+../qe_logic.cpp: qe3_u.c
+	@echo QE_LOGIC
+
+
 id_enclave_u.c: $(SGX_EDGER8R) ../id_enclave/id_enclave.edl
	$(SGX_EDGER8R) --untrusted ../id_enclave/id_enclave.edl
	@echo "GEN  =>  $@"
